Push to production (new script: tag_and_deploy)

shahdyousefak opened the pull request:
scripts/tag_and_deploy

- Parses all GitHub workflow files to figure out what components need versioned tags.
- Identifies the most recent tag for each component (by directory).
- Creates a new tag with an incremented patch version (e.g. 1.2.3 → 1.2.4).
- Pushes that tag to GitHub, triggering the corresponding GitHub Action workflow for that component.

You can run it either:

- with no args -- tags everything (preprocessors, handlers, services, orchestrator)
- with a specific path, e.g., ./scripts/tag_and_deploy services/tat, to tag only that component.

The script's output is logged in /var/docker/image/testing/tag-and-deploy, which is adjacent to how the rest of our scripts log.

Sample output:

```
shahdy@unicorn:~/IMAGE-server$ ./scripts/tag_and_deploy services/tat
Updating main branch...
From github.com:Shared-Reality-Lab/IMAGE-server
 * branch              main       -> FETCH_HEAD
Already up to date.
Parsing GitHub workflow tag patterns...
Tagging services...
Explicit services provided: services/tat
Tagging service-tat: 0.6.8 -> 0.6.9 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           service-tat-0.6.9 -> service-tat-0.6.9
Done. Check workflows at:
  https://github.com/Shared-Reality-Lab/IMAGE-server/actions
Please note, to complete deployment to Pegasus, run imageup after the above workflows complete! :)
```

```
shahdy@unicorn:~/IMAGE-server$ ./scripts/tag_and_deploy 
shahdy@unicorn:/var/docker/image/testing/tag-and-deploy$ cat tag_and_deploy.2025.08.01-10.07.35
Updating main branch...
From github.com:Shared-Reality-Lab/IMAGE-server
 * branch              main       -> FETCH_HEAD
Already up to date.
Parsing GitHub workflow tag patterns...
Tagging components...
Tagging orchestrator: 0.4.4 -> 0.4.5 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           orchestrator-0.4.5 -> orchestrator-0.4.5
Tagging preprocessor-autour: 0.9.0 -> 0.9.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-autour-0.9.1 -> preprocessor-autour-0.9.1
No existing tags for preprocessor-celebrity-detector, skipping.
No existing tags for preprocessor-clothes-detector, skipping.
Tagging preprocessor-collage-detector: 0.6.0 -> 0.6.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-collage-detector-0.6.1 -> preprocessor-collage-detector-0.6.1
Tagging preprocessor-content-categoriser: 1.1.0 -> 1.1.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-content-categoriser-1.1.1 -> preprocessor-content-categoriser-1.1.1
Tagging preprocessor-depth-map-generator: 0.8.0 -> 0.8.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-depth-map-generator-0.8.1 -> preprocessor-depth-map-generator-0.8.1
Tagging preprocessor-graphic-caption: 1.1.0 -> 1.1.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-graphic-caption-1.1.1 -> preprocessor-graphic-caption-1.1.1
Tagging preprocessor-graphic-tagger: 1.1.0 -> 1.1.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-graphic-tagger-1.1.1 -> preprocessor-graphic-tagger-1.1.1
Tagging preprocessor-grouping: 1.1.0 -> 1.1.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-grouping-1.1.1 -> preprocessor-grouping-1.1.1
Skipping preprocessors/hello-preprocessor: no matching workflow tag
Tagging preprocessor-line-charts: 0.1.0 -> 0.1.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-line-charts-0.1.1 -> preprocessor-line-charts-0.1.1
Tagging preprocessor-mmsemantic-segmentation: 1.0.0 -> 1.0.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-mmsemantic-segmentation-1.0.1 -> preprocessor-mmsemantic-segmentation-1.0.1
Tagging preprocessor-multistage-diagram-segmentation: 0.4.0 -> 0.4.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-multistage-diagram-segmentation-0.4.1 -> preprocessor-multistage-diagram-segmentation-0.4.1
Tagging preprocessor-ner: 0.1.0 -> 0.1.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-ner-0.1.1 -> preprocessor-ner-0.1.1
Tagging preprocessor-nominatim: 1.0.0 -> 1.0.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-nominatim-1.0.1 -> preprocessor-nominatim-1.0.1
Tagging preprocessor-object-depth-calculator: 0.2.0 -> 0.2.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-object-depth-calculator-0.2.1 -> preprocessor-object-depth-calculator-0.2.1
Tagging preprocessor-object-detection-azure: 0.2.0 -> 0.2.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-object-detection-azure-0.2.1 -> preprocessor-object-detection-azure-0.2.1
Tagging preprocessor-ocr-clouds: 1.0.0 -> 1.0.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-ocr-clouds-1.0.1 -> preprocessor-ocr-clouds-1.0.1
Tagging preprocessor-openstreetmap: 1.1.0 -> 1.1.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-openstreetmap-1.1.1 -> preprocessor-openstreetmap-1.1.1
Tagging preprocessor-resize-graphic: 0.2.0 -> 0.2.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-resize-graphic-0.2.1 -> preprocessor-resize-graphic-0.2.1
Tagging preprocessor-sorting: 1.1.0 -> 1.1.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-sorting-1.1.1 -> preprocessor-sorting-1.1.1
Tagging preprocessor-text-followup: 0.2.0 -> 0.2.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-text-followup-0.2.1 -> preprocessor-text-followup-0.2.1
Tagging preprocessor-object-detection-yolo: 0.2.0 -> 0.2.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           preprocessor-object-detection-yolo-0.2.1 -> preprocessor-object-detection-yolo-0.2.1
Tagging handler-autour: 0.1.0 -> 0.1.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-autour-0.1.1 -> handler-autour-0.1.1
Skipping handlers/hello-handler: no matching workflow tag
Skipping handlers/hello-svg-handler: no matching workflow tag
Tagging handler-high-charts: 0.8.6 -> 0.8.7 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-high-charts-0.8.7 -> handler-high-charts-0.8.7
Tagging handler-map-tactile-svg: 0.3.0 -> 0.3.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-map-tactile-svg-0.3.1 -> handler-map-tactile-svg-0.3.1
Tagging handler-motd: 0.0.7 -> 0.0.8 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-motd-0.0.8 -> handler-motd-0.0.8
Tagging handler-multistage-diagram-tactile-svg: 0.2.6 -> 0.2.7 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-multistage-diagram-tactile-svg-0.2.7 -> handler-multistage-diagram-tactile-svg-0.2.7
Tagging handler-ocr: 0.0.7 -> 0.0.8 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-ocr-0.0.8 -> handler-ocr-0.0.8
Tagging handler-osm-streets: 0.0.7 -> 0.0.8 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-osm-streets-0.0.8 -> handler-osm-streets-0.0.8
Tagging handler-photo-audio: 0.1.6 -> 0.1.7 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-photo-audio-0.1.7 -> handler-photo-audio-0.1.7
Tagging handler-photo-audio-haptics: 0.1.6 -> 0.1.7 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-photo-audio-haptics-0.1.7 -> handler-photo-audio-haptics-0.1.7
Tagging handler-photo-tactile-svg: 0.3.0 -> 0.3.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-photo-tactile-svg-0.3.1 -> handler-photo-tactile-svg-0.3.1
Tagging handler-svg-depth-map: 0.8.6 -> 0.8.7 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-svg-depth-map-0.8.7 -> handler-svg-depth-map-0.8.7
Tagging handler-svg-od: 0.6.7 -> 0.6.8 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-svg-od-0.6.8 -> handler-svg-od-0.6.8
Tagging handler-svg-open-street-map: 0.0.2 -> 0.0.3 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-svg-open-street-map-0.0.3 -> handler-svg-open-street-map-0.0.3
Tagging handler-svg-semantic-seg: 0.7.7 -> 0.7.8 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-svg-semantic-seg-0.7.8 -> handler-svg-semantic-seg-0.7.8
Tagging handler-text-followup: 0.1.0 -> 0.1.1 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-text-followup-0.1.1 -> handler-text-followup-0.1.1
Tagging service-espnet-tts: 0.2.6 -> 0.2.7 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           service-espnet-tts-0.2.7 -> service-espnet-tts-0.2.7
Tagging service-espnet-tts-fr: 0.2.6 -> 0.2.7 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           service-espnet-tts-fr-0.2.7 -> service-espnet-tts-fr-0.2.7
Tagging service-monarch-link-app: 0.2.6 -> 0.2.7 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           service-monarch-link-app-0.2.7 -> service-monarch-link-app-0.2.7
Tagging service-translation: 0.6.6 -> 0.6.7 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           service-translation-0.6.7 -> service-translation-0.6.7
Skipping services/supercollider-images: no matching workflow tag
Tagging service-tat: 0.6.7 -> 0.6.8 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           service-tat-0.6.8 -> service-tat-0.6.8
Done. Check workflows at:
  https://github.com/Shared-Reality-Lab/IMAGE-server/actions
Please note, to complete deployment to Pegasus, run imageup after the above workflows complete! :)


```

Please ensure you've followed the checklist and provide all the required information *before* requesting a review.
If you do not have everything applicable to your PR, it will not be reviewed!
If you don't know what something is or if it applies to you, ask!

Please note that PRs from external contributors who have not agreed to [our Contributor License Agreement](/CLA.md) will not be considered.
To accept it, include `I agree to the [current Contributor License Agreement](/CLA.md)` in this pull request.

Don't delete below this line.

---

## Required Information

- [X] I referenced the issue addressed in this PR.
- [X] I described the changes made and how these address the issue.
- [X] I described how I tested these changes.

## Coding/Commit Requirements

* [X] I followed applicable coding standards where appropriate (e.g., [PEP8](https://pep8.org/))
* [X] I have not committed any models or other large files.

## New Component Checklist (**mandatory** for new microservices)

* [ ] I added an entry to `docker-compose.yml` and `build.yml`.
* [ ] I created A CI workflow under `.github/workflows`.
* [ ] I have created a `README.md` file that describes what the component does and what it depends on (other microservices, ML models, etc.).

OR
* [X] I have not added a new component in this PR.

shahdyousefak committed:
broken tag parsing logic

shahdyousefak committed:
Parses .github/workflows/*.yml files to extract tag patterns + directory paths + maps components  to tag prefixes

shahdyousefak committed:
see which components were actually bumped

shahdyousefak committed:
bug fix: now + TAG_AND_DEPLOY_ENABLED_DIRS=preprocessors handlers services orchestrator

shahdyousefak committed:
adding logging block

shahdyousefak committed:
bug fix: orchestrator tagging

shahdyousefak committed:
Bug fix: orchestrator tagging

shahdyousefak committed:
Adding a cascading incremental system to tag versioning

shahdyousefak committed:
supporting service-specific deployment

shahdyousefak committed:
removing log.txt

shahdyousefak committed:
Correcting comment

jeffbl commented on scripts/tag_and_deploy:
```diff
@@ -0,0 +1,164 @@
+# Tags IMAGE-server components that match GitHub Actions workflows with tag triggers
+# Usage:
+#	./scripts/tag_and_deploy # tags all services
+#	./scripts/tag_and_deploy services/tat # tags 1+ services
```
This means that it can take more than one argument for explicit services, correct?

jeffbl commented on scripts/tag_and_deploy:
```diff
@@ -0,0 +1,164 @@
+
+echo "Tagging services..."
+
+if [ "${#EXPLICIT_SERVICES[@]}" -gt 0 ]; then
```
Add comment that this block is for the case of user manually specifying specific components. Confusing since "services" is not generic - it has a specific meaning in IMAGE, so maybe EXPLICIT_COMPONENTS would be more clear?

jeffbl reviewed (changes requested):
Nice that this can do just one or two, or also just do all in one pass. Major issue is the way the version tag is incremented, which must be fixed before we use this.

shahdyousefak commented on scripts/tag_and_deploy:
```diff
@@ -0,0 +1,164 @@
+  PATCH=$(echo "$VERSION" | cut -d. -f3)
+
+
+  # cascading incrementing from the latest existing tag
```
https://semver.org/ ? 
another idea,  we can keep auto-patch bump, but remove cascading rollover. If a component needs a major/minor bump, we can have that be done manually via argument or commit tag?
but note: not all components are on the same version number (some components were tagged individually more than others, some components are newer, etc..)

shahdyousefak replied:
we dont really have a philosophy in place for what type of tag bump a component should obtain, its been very arbitrary. open to having this discussion as a part of this work item.

shahdyousefak committed:
renaming to EXPLICIT_COMPONENTS

shahdyousefak committed:
Modified tag bumping to default to patch, with version bump flag support for minor and major

shahdyousefak commented:
Updated the versioning to default to patch bump by default, and introduced flag versioning:-
--bump-patch (default)
--bump-minor
--bump-major

Eg: shahdy@unicorn:~/IMAGE-server$ ./scripts/tag_and_deploy --bump-minor services/tat handlers/multistage-diagram-tactile-svg
Updating main branch...

Explicit components provided: services/tat handlers/multistage-diagram-tactile-svg
Tagging service-tat: 0.6.9 -> 0.7.0 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           service-tat-0.7.0 -> service-tat-0.7.0
Tagging handler-multistage-diagram-tactile-svg: 0.2.7 -> 0.3.0 at HEAD
To github.com:Shared-Reality-Lab/IMAGE-server.git
 * [new tag]           handler-multistage-diagram-tactile-svg-0.3.0 -> handler-multistage-diagram-tactile-svg-0.3.0
Done. Check workflows at:
  https://github.com/Shared-Reality-Lab/IMAGE-server/actions
Please note, to complete deployment to Pegasus, run imageup after the above workflows complete! :)

shahdyousefak committed:
initial version 0.0.1 when no existing tags are found
